### ISSUE TITLE
Rework user_index/ptr

### ISSUE THREAD

**Issue by tjhei** [dealii/dealii#553]
There are several problems with user_index and user_ptr:
  - user_index on faces is not retained after mesh refinement, while it is for cells
  - internal algorithms have to save/restore user_indices even if they are not used by the user
  - not allowing arbitrary data is limiting
  - the exclusive choice between ptr and index is confusing

I would prefer an external object (maybe called TriangulationUserData<T,dim>):
  - this attaches an object of type T to all objects of dimension dim (cells, faces, etc.)
  - you can have arbitrary many of those objects
  - we can provide SolutionTransfer-like support (callback for "interpolation"/"restriction")
  - we can support distributed computations

Internally TriangulationUserData<T> might have a std::vector<T> and an IndexSet or a map or something.

Thoughts?

**Comment by bangerth** [comment]
Yes, I agree. I could imagine an interface of the kind
```
template <typename Container, typename Value>
class TriangulationUserData {
  public:
    Value & operator[] (typename Container::cell_iterator);
    Value & operator[] (typename Container::face_iterator);
};
```
It could internally use two `std::map<typename Container::{cell,face}_iterator, Value>` objects.

**Comment by tjhei** [comment]
I think cell/face/vertex data should be separate (that is why I included a ```dim``` parameter). I agree with the ```operator []`` syntax, but the internal storage mustn't depend on the iterators to persist between refinements.

**Comment by bangerth** [comment]
Fair enough. One could use the `CellIndex` you wrote (assuming a way can be found to re-create a cell from its `CellIndex`. The same could be done with face iterators.

**Comment by guidokanschat** [comment]
I am wondering: how does the conversion from an index to a cell position and back work in p4est? Maybe we can learn how to do this without another vector?

**Comment by bangerth** [comment]
In p4est, they are the same: `CellIndex==cell_iterator`.

**Comment by tjhei** [comment]
The problems with ```CellIndex``` are:
  - finding the cell iterator takes log(level) operations 
  - querying the ```CellIndex``` takes log(level) operations (we could precompute them)
  - the representation is not efficient (we use a std::vector so we can support anisotropic refinement)

> In p4est, they are the same: CellIndex==cell_iterator.

Really? I thought p4est stores a flat array of active quadrants and their id but you can not directly jump from an id to the quadrant.

**Comment by bangerth** [comment]
Yes, it's true that access to the data in the array given a quadrant/CellIndex requires log(N) effort (namely a search for an element in a sorted, linear array using something like bisection search).
